### PR TITLE
Fix clip copy boundaries

### PR DESCRIPTION
**Pull request by vsverchinsky**
Resolves: #4134

- [x] I signed [CLA](https://www.audacityteam.org/cla/)
- [x] The title of the pull request describes an issue it addresses
- [x] If changes are extensive, then there is a sequence of easily reviewable commits
- [x] Each commit's message describes its purpose and effects
- [x] There are no behavior changes unnecessary for the stated purpose of the PR

Recommended:
- [x] Each commit compiles and runs on my machine without known undesirable changes of behavior

### CODE REVIEW

**Comment by Paul-Licameli** [comment]
I understand how this fixes the original report.  I fear it might not be the last bug that is found and fixed, involving floating point durations that are not exact multples of the sample period.  Maybe some future fix will be more comprehensive.

The only objection I have, as I explained, is slight misuse of assertions.  There is a duty to add some comments when you add assertions, and offer some informal proof in comments that the assertions can't fail at least now when you add them.

**Review by Paul-Licameli — CHANGES_REQUESTED** [review]
Better assertion discipline, that's all I'd like to see

File: src/WaveClip.cpp [hunk 1]
```diff
@@ -83,20 +83,36 @@ WaveClip::WaveClip(const WaveClip& orig,
                    bool copyCutlines,
                    double t0, double t1)
 {
+   assert(orig.CountSamples(t0, t1) >= 1);
```
[the remark this answers is not in the document]
Comment (vsverchinsky) [reply src/WaveClip.cpp:86]: > you must reexamine all uses of this constructor and prove that the assertion violation can't happen.

I did that

> If you add this assertion, you must add a @pre comment in the header file ...

I don't understand, do you mean that `assert` is essentially useless and we should use `@pre` instead?

[the remark this answers is not in the document]
Comment (Paul-Licameli) [reply src/WaveClip.cpp:86]: Not instead, but also. I mean that an assertion at the top of the function means the function assumes a precondition, and that should be part of its documentation, at least as a doxygen comment in the header file.

**Review by Paul-Licameli — APPROVED** [review]

File: src/WaveClip.cpp [hunk 2]
```diff
@@ -83,20 +83,36 @@ WaveClip::WaveClip(const WaveClip& orig,
...
    mColourIndex = orig.mColourIndex;
 
    mIsPlaceholder = orig.GetIsPlaceholder();
 
-   auto s0 = orig.TimeToSequenceSamples(t0);
-   auto s1 = orig.TimeToSequenceSamples(t1);
-
    mSequence = std::make_unique<Sequence>(*orig.mSequence, factory);
 
+   assert(GetPlaySamplesCount() > 0);
```
[the remark this answers is not in the document]
Comment (vsverchinsky) [reply src/WaveClip.cpp]: > nothing really bad happens except that you make an empty clip.

True, that means actually that requirement `orig.CountSamples(t0, t1) > 0` isn't necessary, it's enough to make sure that empty clips can't appear on the track, as they are causing troubles.

**Pull request merged**